File: caller/bold-mesh-python/src/bold/mesh/blocks.py
"""Content-block builders -- match the wire contract exactly
(server/internal/gateway/contract.go's contentBlockRequest): each block is
{"type", "text"?, "media_type"?, "data"?, "url"?}.

Every builder besides text_block() accepts path/data/media_type/url as
either a single value or a list. Passing a list on any one of them builds
several blocks at once and returns a list, e.g.:

    image_block(path=["photo.jpg", "ninja.jpg"])
    -> [ {...block for photo.jpg...}, {...block for ninja.jpg...} ]

so it can be spliced straight into a content=[...] list -- llm_caller
flattens one level of nested lists in `content` before sending, so mixing
scalar and batch blocks in the same call just works:

    content=[
        text_block("Compare these"),
        image_block(path=["photo.jpg", "ninja.jpg"]),
    ]
"""

import base64
import mimetypes
from typing import List, Optional, Union

Scalar = Union[str, bytes]
OneOrMany = Union[Scalar, List[Scalar]]


def text_block(text: str) -> dict:
    return {"type": "text", "text": text}


def _as_list(value):
    if value is None:
        return None
    return value if isinstance(value, list) else [value]


def _one_block(
    block_type: str,
    path: Optional[str],
    data: Optional[bytes],
    media_type: Optional[str],
    url: Optional[str],
    default_media_type: Optional[str] = None,
) -> dict:
    sources = sum(x is not None for x in (path, data, url))
    if sources != 1:
        raise ValueError(f"{block_type} block needs exactly one of path, data, or url")

    if url is not None:
        return {"type": block_type, "url": url}

    if path is not None:
        guessed, _ = mimetypes.guess_type(path)
        # An explicit media_type always wins; otherwise prefer what
        # mimetypes actually detected from the file over a generic
        # per-block-type default (e.g. a .docx passed to document_block
        # must not get force-tagged as application/pdf).
        resolved_media_type = media_type or guessed or default_media_type
        if not resolved_media_type:
            raise ValueError(f"could not guess a media type for {path!r} -- pass media_type explicitly")
        with open(path, "rb") as f:
            data = f.read()
        media_type = resolved_media_type
    elif not media_type:
        media_type = default_media_type

    if not media_type:
        raise ValueError(f"{block_type} block with inline data needs media_type")

    return {
        "type": block_type,
        "media_type": media_type,
        "data": base64.standard_b64encode(data).decode("ascii"),
    }


def _build_blocks(block_type: str, *, path=None, data=None, media_type=None, url=None, default_media_type=None):
    is_batch = isinstance(path, list) or isinstance(data, list) or isinstance(url, list)

    paths = _as_list(path)
    datas = _as_list(data)
    urls = _as_list(url)
    media_types = media_type if isinstance(media_type, list) else None

    n = len(paths or datas or urls or [None])
    for name, seq in (("path", paths), ("data", datas), ("url", urls), ("media_type", media_types)):
        if seq is not None and len(seq) != n:
            raise ValueError(f"{block_type} block: {name} list length ({len(seq)}) doesn't match the others ({n})")

    blocks = [
        _one_block(
            block_type,
            paths[i] if paths else None,
            datas[i] if datas else None,
            media_types[i] if media_types else media_type,
            urls[i] if urls else None,
            default_media_type=default_media_type,
        )
        for i in range(n)
    ]
    return blocks if is_batch else blocks[0]


def image_block(path: OneOrMany = None, data: OneOrMany = None, media_type: OneOrMany = None, url: OneOrMany = None):
    """An image content block (or list of blocks, if any argument is a list)."""
    return _build_blocks("image", path=path, data=data, media_type=media_type, url=url)


def document_block(path: OneOrMany = None, data: OneOrMany = None, media_type: OneOrMany = None, url: OneOrMany = None):
    """A document content block (falls back to application/pdf only when
    mimetypes can't identify the file and no media_type was given)."""
    return _build_blocks("document", path=path, data=data, media_type=media_type, url=url, default_media_type="application/pdf")


def audio_block(path: OneOrMany = None, data: OneOrMany = None, media_type: OneOrMany = None, url: OneOrMany = None):
    return _build_blocks("audio", path=path, data=data, media_type=media_type, url=url)


def video_block(path: OneOrMany = None, data: OneOrMany = None, media_type: OneOrMany = None, url: OneOrMany = None):
    return _build_blocks("video", path=path, data=data, media_type=media_type, url=url)


def flatten_content(content: list) -> list:
    """Flattens one level of nesting -- lets a batch block builder's list
    output sit directly inside a content=[...] list alongside scalar
    blocks."""
    flat = []
    for item in content:
        if isinstance(item, list):
            flat.extend(item)
        else:
            flat.append(item)
    return flat

File: caller/bold-mesh-typescript/src/blocks.ts
/**
 * Content-block builders -- match the wire contract exactly
 * (server/internal/gateway/contract.go's contentBlockRequest): each block
 * is { type, text?, media_type?, data?, url? }.
 *
 * Every builder besides textBlock() accepts path/data/mediaType/url as
 * either a single value or an array. Passing an array on any one of them
 * builds several blocks at once and returns an array, e.g.:
 *
 *   imageBlock({ path: ["photo.jpg", "ninja.jpg"] })
 *   -> [ {...block for photo.jpg...}, {...block for ninja.jpg...} ]
 *
 * so it can be spliced straight into a content: [...] array -- llmCaller
 * flattens one level of nested arrays in `content` before sending, so
 * mixing scalar and batch blocks in the same call just works:
 *
 *   content: [
 *     textBlock("Compare these"),
 *     imageBlock({ path: ["photo.jpg", "ninja.jpg"] }),
 *   ]
 */

import type { ContentBlock, ContentBlockType } from "./types";

export function textBlock(text: string): ContentBlock {
  return { type: "text", text };
}

function asArray<T>(value: T | T[] | undefined): T[] | undefined {
  if (value === undefined) return undefined;
  return Array.isArray(value) ? value : [value];
}

function base64Encode(bytes: Uint8Array): string {
  if (typeof Buffer !== "undefined") {
    return Buffer.from(bytes).toString("base64"); // Node
  }
  let binary = "";
  for (const b of bytes) binary += String.fromCharCode(b);
  return btoa(binary); // browser
}

function guessMediaType(path: string): string | undefined {
  const ext = path.slice(path.lastIndexOf(".") + 1).toLowerCase();
  const table: Record<string, string> = {
    jpg: "image/jpeg", jpeg: "image/jpeg", png: "image/png", gif: "image/gif",
    webp: "image/webp", bmp: "image/bmp", svg: "image/svg+xml", heic: "image/heic",
    pdf: "application/pdf", doc: "application/msword", txt: "text/plain",
    md: "text/markdown", csv: "text/csv", json: "application/json",
    docx: "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    xlsx: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    pptx: "application/vnd.openxmlformats-officedocument.presentationml.presentation",
    mp3: "audio/mpeg", wav: "audio/wav", ogg: "audio/ogg", m4a: "audio/mp4",
    flac: "audio/flac", aac: "audio/aac",
    mp4: "video/mp4", mov: "video/quicktime", webm: "video/webm", mpeg: "video/mpeg",
  };
  return table[ext];
}

async function readFile(path: string): Promise<Uint8Array> {
  const fs = await import("fs/promises");
  return new Uint8Array(await fs.readFile(path));
}

interface BlockOpts {
  path?: string | string[];
  data?: Uint8Array | Uint8Array[];
  mediaType?: string | string[];
  url?: string | string[];
}

async function oneBlock(
  type: ContentBlockType,
  path: string | undefined,
  data: Uint8Array | undefined,
  mediaType: string | undefined,
  url: string | undefined,
  defaultMediaType?: string,
): Promise<ContentBlock> {
  const sources = [path, data, url].filter((x) => x !== undefined).length;
  if (sources !== 1) {
    throw new Error(`${type} block needs exactly one of path, data, or url`);
  }

  if (url !== undefined) {
    return { type, url };
  }

  let resolvedMediaType = mediaType;
  let resolvedData = data;
  if (path !== undefined) {
    const guessed = guessMediaType(path);
    resolvedMediaType = mediaType || guessed || defaultMediaType;
    if (!resolvedMediaType) {
      throw new Error(`could not guess a media type for ${path} -- pass mediaType explicitly`);
    }
    resolvedData = await readFile(path);
  } else if (!resolvedMediaType) {
    resolvedMediaType = defaultMediaType;
  }

  if (!resolvedMediaType) {
    throw new Error(`${type} block with inline data needs mediaType`);
  }

  return { type, media_type: resolvedMediaType, data: base64Encode(resolvedData!) };
}

async function buildBlocks(
  type: ContentBlockType,
  opts: BlockOpts,
  defaultMediaType?: string,
): Promise<ContentBlock | ContentBlock[]> {
  const isBatch = Array.isArray(opts.path) || Array.isArray(opts.data) || Array.isArray(opts.url);

  const paths = asArray(opts.path);
  const datas = asArray(opts.data);
  const urls = asArray(opts.url);
  const mediaTypes = Array.isArray(opts.mediaType) ? opts.mediaType : undefined;

  const n = (paths ?? datas ?? urls ?? [undefined]).length;
  for (const [name, seq] of [["path", paths], ["data", datas], ["url", urls], ["mediaType", mediaTypes]] as const) {
    if (seq !== undefined && seq.length !== n) {
      throw new Error(`${type} block: ${name} array length (${seq.length}) doesn't match the others (${n})`);
    }
  }

  const blocks = await Promise.all(
    Array.from({ length: n }, (_, i) =>
      oneBlock(
        type,
        paths?.[i],
        datas?.[i],
        mediaTypes ? mediaTypes[i] : (opts.mediaType as string | undefined),
        urls?.[i],
        defaultMediaType,
      ),
    ),
  );
  return isBatch ? blocks : blocks[0];
}

/** An image content block (or array of blocks, if any option is an array). */
export function imageBlock(opts: BlockOpts): Promise<ContentBlock | ContentBlock[]> {
  return buildBlocks("image", opts);
}

/** A document content block (falls back to application/pdf only when the
 * extension can't be identified and no mediaType was given). */
export function documentBlock(opts: BlockOpts): Promise<ContentBlock | ContentBlock[]> {
  return buildBlocks("document", opts, "application/pdf");
}

export function audioBlock(opts: BlockOpts): Promise<ContentBlock | ContentBlock[]> {
  return buildBlocks("audio", opts);
}

export function videoBlock(opts: BlockOpts): Promise<ContentBlock | ContentBlock[]> {
  return buildBlocks("video", opts);
}

/** Flattens one level of nesting -- lets a batch block builder's array
 * output sit directly inside a content: [...] array alongside scalar
 * blocks. */
export function flattenContent(content: (ContentBlock | ContentBlock[])[]): ContentBlock[] {
  const flat: ContentBlock[] = [];
  for (const item of content) {
    if (Array.isArray(item)) flat.push(...item);
    else flat.push(item);
  }
  return flat;
}
